src/model.js

```diff
@@ -0,0 +1,187 @@
+
+const MAX_PAGE_SIZE = 100; // maximum number of results returned from portal per request
+const FIELD_DICTIONARY = require('./helpers/field-dictionary');
+const PORTAL_URL = 'http://www.arcgis.com/sharing/rest/search';
+const PROJ4_WKIDS = [4326, 4269, 3857, 3785, 900913, 102113]; // supported wkids by proj4
+
+class ArcgisSearchModel {
+  // Main getData method which is used to send data to Koop 
+  async getData(req, callback) {
+    try {
+      const portalQuery = this.buildPortalQuery(req.query);
+      const items = await this.getPortalItems(PORTAL_URL, portalQuery, MAX_PAGE_SIZE);
+      const geojson = this.getGeoJson(items, FIELD_DICTIONARY);
+      callback(null, geojson);
+    } catch (error) {
+      throw new ArcgisSearchProviderError(
+        error?.message || 'Error in Arcgis Search Provider',
+        error?.statusCode || 500
+      );
+    }
+  }
+
+  buildPortalQuery(requestQuery) {
+    const portalQuery = { f: 'json' };
+    // TODO: ensure appropriate quotes for Search in Online
+    // Remove any variant of 1=1 as it is not recognized by AGO search; replace = with; replace and ' with "
+    portalQuery.q =
+      (requestQuery.where || '*')
+        .replace(/1=1|(\(1=1\))|(AND\s1=1)|(AND\s\(1=1\))/g, '')
+        .replace(/\s+=\s+/g, ':')
+        .replace(/'/g, '"')
+        .replace(/[ \t]+$/, '');
+
+    portalQuery.num = requestQuery.resultRecordCount || 100;
+    portalQuery.start = requestQuery.resultOffset || 1;
+
+    if (requestQuery.orderByFields) {
+      this.addSortOptionsToPortalQuery(portalQuery, requestQuery.orderByFields);
+    }
+
+    if (this.isSupportedGeometry(requestQuery)) {
+      this.addBboxToPortalQuery(portalQuery, requestQuery);
+    }
+
+    return portalQuery;
+  }
+
+  getSpatialReference(requestQuery) {
+    const { geometry: { spatialReference: { wkid } = {} } } = requestQuery;
+    const queryWkid = wkid || requestQuery.inSR;
+
+    // assume 4326 if no SR is provided
+    if (!queryWkid) {
+      return 4326;
+    }
+
+    // 102100 is the old Esri code for 3857 but not recognized for proj4
+    if (queryWkid === 102100) {
+      return 3857;
+    }
+    // If the input wkid is one of the set known to proj4, return it in an object
+    if (PROJ4_WKIDS.includes(queryWkid)) {
+      return queryWkid;
+    }
+
+    logger.debug(`unsupported wkid ${queryWkid} provided`);
+  }
+
+  isSupportedGeometry(requestQuery) {
+    return (
+      requestQuery.geometry &&
+      requestQuery.geometryType === 'esriGeometryEnvelope'
+    );
+  }
+
+  getExtent(geometry) {
+    if (!objTypeGeometryQuerySchema.validate(geometry).error) {
+      const { xmin, ymin, xmax, ymax } = geometry;
+      return [[xmin, ymin], [xmax, ymax]];
+    }
+
+    const extent = geometry.split(',').map(Number);
+    if (!arrayTypeGeometryQuerySchema.validate(extent).error) {
+      return [extent.splice(0, 2), extent];
+    }
```

Comment:
as noted above, you are validating and parsing in same step here.  Let's separate out validation.

Reply:
I have added a basic validation function. However, functions here are not doing much of a validation but helping construct bbox query for envelope geometry and supported spatial references. We'd still want to send the request to Koop server for other spatial references and geometry types as Geoservices may support it.